### ISSUE TITLE
false positives in `Enum` stub definition

### ISSUE THREAD

**Issue by jorenham** [DetachHead/basedpyright#1378]
### Description

from the [typing spec](https://typing.python.org/en/latest/spec/enums.html#defining-members):

> Within a type stub, members can be defined using the actual runtime values, or a placeholder of `...` can be used:
> 
> ```pyi
> class Pet(Enum):
>     genus: str  # Non-member attribute
>     species: str  # Non-member attribute
> 
>     CAT = 1  # Member attribute with known value and type
>     DOG = cast(int, ...)  # Member attribute with unknown value and known type
>     BIRD = ...  # Member attribute with unknown value and type
> ```

given this example, basedpyright reports:

```
Instance variable "genus" is not initialized in the class body or __init__ method  (reportUninitializedInstanceVariable)
Instance variable "species" is not initialized in the class body or __init__ method  (reportUninitializedInstanceVariable)
Conversion of type `EllipsisType` to type `int` may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to `object` first.  (reportInvalidCast)
```

playgrounds:

- [basedpyright](https://basedpyright.com/?typeCheckingMode=all&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDGAhgM4wBQokUApigK7RY55QCiDE55BANqSSgAFGjAAUHRgEoAXOSgKoaOvRIyoZEAoDEUAHJgUAWgg0IAIxpaiMGCCTn6MGvMUkENAkhpqNdnfqGJmaW1rb2js7cilAAwgCCACpQALxQAIwBALIhVlA2dg5ONFAA7kgwABZQANYoYKUoUABuRLz0JUQoACawiC4xACIA8gDiqYSk4qgwADRQAHRLUtm5YYWRJeVVUPQodQ1Nre2dPbX1jX0ergoAQgCSAEqDE0sLqxZ5BRHFZRXVewOl2OHXyZ3g1yAA)
- [pyright](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDGAhgM4wBQokUApigK7RY55QCiDE55BANqSSgAFGjAAUHRgEoAXOSgKoaOvRIyoZEAoDEUAHJgUAWgg0IAIxpaiMGCCTn6MGvMUkENAkhpqNdnfqGJmaW1rb2js7cilAAwgCCACpQALxQAIwBALIhVlA2dg5ONFAA7kgwABZQANYoYKUoUABuRLz0JUQoACawiC4xACIA8gDiqYSk4qgwADRQAHRLUtm5YYWRJeVVUPQodQ1Nre2dPbX1jX0ergoAQgCSAEqDE0sLqxZ5BRHFZRXVewOl2OHXyZ3g1yAA) (no errors)

**Comment by KotlinIsland** [comment]
wow, this is pretty serious. I wonder why the primer and tests never picked these up

**Comment by jorenham** [comment]
> wow, this is pretty serious. I wonder why the primer and tests never picked these up

Maybe because enums suck in Python so nobody uses them 🤷?

**Comment by DetachHead** [comment]
the `reportUninitializedInstanceVariable` errors happen in pyright as well, the rule just isn't enabled by default in strict mode. [playground](https://pyright-play.net/?strict=true&reportUninitializedInstanceVariable=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDGAhgM4wBQokUApigK7RY55QCiDE55BANqSSgAFGjAAUHRgEoAXOSgKoaOvRIyoZEAoDEUAHJgUAWgg0IAIxpaiMGCCTn6MGvMUkENAkhpqNdnfqGJmaW1rb2js7cilAAwgCCACpQALxQAIwBALIhVlA2dg5ONFAA7kgwABZQANYoYKUoUABuRLz0JUQoACawiC4xACIA8gDiqYSk4qgwADRQAHRLUtm5YYWRJeVVUPQodQ1Nre2dPbX1jX0ergoAQgCSAEqDE0sLqxZ5BRHFZRXVewOl2OHXyZ3g1yAA)

i think the error is correct though, because you'll get a runtime error if you try to use it without assigning a value:

```py
class Pet(Enum):
    genus: str  # Non-member attribute
    species: str  # Non-member attribute

print(Pet.genus) # AttributeError: type object 'Pet' has no attribute 'genus'
```

i noticed it says "Within a type stub". if you do this in a `.pyi` file the error is not reported (note that the playground assumes your code is in a `.py` file, see https://github.com/DetachHead/basedpyright-playground/issues/11).

as for the `reportInvalidCast` error (this rule is unique to basedpyright), i guess casting an ellipsis on an enum value in a `.pyi` file should be special cased to not report this error.

**Comment by jorenham** [comment]
Ok makes sense 👌
